### PR TITLE
lodash: Add clone(Deep)WithCustomizer parameters

### PR DESCRIPTION
**Pull request by mikejhill**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present.
  * _***Directory v3 has two issues flagged, but those do not relate to this PR**_

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://lodash.com/docs/4.17.4#cloneWith>>
- [x] Increase the version number in the header if appropriate.
  * _***No version number increment necessary**_
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.
  * _***tslint.json already exists**_

### CODE REVIEW

File: lodash/index.d.ts [hunk 1]
```diff
@@ -11202,7 +11202,7 @@ declare namespace _ {
     }
 
     //_.cloneDeepWith
-    type CloneDeepWithCustomizer<TValue, TResult> = (value: TValue) => TResult;
+    type CloneDeepWithCustomizer<TValue, TResult> = (value: TValue, key?: number|string, object?: any, stack?: any) => TResult;
```
Comment (mhegazy) [inline lodash/index.d.ts]: do not use `?` in callbacks. lodash will always call it with 4 arguments, functions are allowed to ignore arguments.

Comment (hill0826) [inline lodash/index.d.ts]: The first internal call to the customizer doesn't include the 2nd, 3rd, or 4th arguments. Only the recursive calls to `baseClone` seem to include the remaining arguments.

We don't care about this when we're creating the customizers as library users, but shouldn't we take this into account for correctness? Or am I misinterpreting the use of optional parameters? Still a bit new to TypeScript.

Comment (hill0826) [inline lodash/index.d.ts]: I don't have a problem making optional parameters required if there's a good case for it. Just seems like this is the right case to use optional parameters.

Want to get this in so we don't have to keep modifying the library source for TS compilation to succeed.

**TS Coding Standards**
http://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#optional-parameters-in-callbacks
> This has a very specific meaning: the done callback might be invoked with 1 argument or might be invoked with 2 arguments.

**Lodash API**
https://lodash.com/docs/4.17.4#cloneWith
https://lodash.com/docs/4.17.4#cloneDeepWith
> The customizer is invoked with up to four arguments; (value [, index|key, object, stack]).

**Lodash Source, as of latest commit**
https://github.com/lodash/lodash/blob/165572022d58d3e39e928b33ff9b9589cc60b67b/dist/lodash.js#L2648 (line 2648)
> result = object ? customizer(value, key, object, stack) : customizer(value);

Comment (andy-ms) [inline lodash/index.d.ts]: This is a proper use then. Usually that is mistaken but in this case it is correct.

**Pull request merged**